src/v/datalake/schema_avro.cc

```diff
@@ -0,0 +1,304 @@
+        return iceberg::int_type{};
+    case avro::AVRO_ARRAY: {
+        if (node->leaves() != 1) {
+            return schema_conversion_exception(fmt::format(
+              "Invalid number of leaves {} in AVRO_ARRAY, expected to have "
+              "only 1 leaf",
+              node->leaves()));
+        }
+        auto array_id = state.next_field_id();
+        auto element_type = node->leafAt(0);
+        auto field_res = field_type_from_avro(element_type, field_name, state);
+        if (field_res.has_error()) {
+            return field_res.error();
+        }
+        if (!field_res.value().has_value()) {
+            return schema_conversion_exception(
+              fmt::format("Unsupported type of AVRO_NULL as an array element"));
+        }
+        return iceberg::list_type::create(
+          array_id,
+          iceberg::field_required::yes,
+          std::move(*field_res.value()));
+    }
+    case avro::AVRO_MAP: {
+        if (node->leaves() != 2) {
+            return schema_conversion_exception(fmt::format(
+              "Invalid number of leaves {} in AVRO_MAP, expected to have "
+              "exactly 2 leaves",
+              node->leaves()));
+        }
+
+        if (node->leafAt(0)->type() != avro::AVRO_STRING) {
+            return schema_conversion_exception(fmt::format(
+              "AVRO_MAP is expected to be a string. Current key type",
+              node->leafAt(0)->type()));
+        }
+        auto value_t_result = field_type_from_avro(
+          node->leafAt(1), field_name, state);
+        if (value_t_result.has_error()) {
+            return value_t_result.error();
+        }
+        if (!value_t_result.value().has_value()) {
+            return schema_conversion_exception(
+              fmt::format("Unsupported type of AVRO_NULL as a map value"));
+        }
+
+        return iceberg::map_type::create(
+          state.next_field_id(),
+          iceberg::string_type{},
+          state.next_field_id(),
+          iceberg::field_required::yes,
+          std::move(*value_t_result.value()));
+    }
+    case avro::AVRO_UNION: {
+        // Avro union is flattened as a struct with fields that are not
+        // required, only one of the fields will be present at a time.
+        iceberg::struct_type ret;
+        ret.fields.reserve(node->leaves());
+        for (size_t i = 0; i < node->leaves(); ++i) {
+            auto leaf = node->leafAt(i);
+            auto result = field_type_from_avro(leaf, field_name, state);
+            if (result.has_error()) {
+                return result.error();
+            }
+            auto field = std::move(result.value());
+            if (field.has_value()) {
+                ret.fields.push_back(iceberg::nested_field::create(
+                  state.next_field_id(),
+                  // union struct fields name are represented as
+                  // <union_name>_<leaf_idx>
+                  fmt::format("{}_{}", field_name, i),
```

Comment:
does this repeat the name twice? Did we decide to just do the index here or not?

Reply:
it is a name of a field in a record plus index of leaf in the union

Reply:
Just to be clear if I have a union of `int` and `string` in that order, how would I query it? 

1. `SELECT my_union.0 FROM redpanda_topic` 
2. `SELECT my_union.my_union_0 FROM redpanda_topic`

Reply:
`SELECT my_union_0 FROM redpanda_topic`

Reply:
Are we worried about conflicts in naming there? You could in theory have a field called `my_union_0` and another field called `my_union` and they could conflict right? That's why I was assuming we'd nest this in a struct because then we wouldn't have conflicts.

Reply:
This is how the JSON encoding for avro works (but they use type names instead of indexes: https://avro.apache.org/docs/1.11.1/specification/#json-encoding)

Reply:
I validated that indeed we will have
```
SELECT my_union.my_union_0 FROM redpanda_topic
```
I will go back to previous version where it will be: 
```
 SELECT my_union.union_0 FROM redpanda_topic
```

Reply:
or maybe `my_union.union_opt_0` to express that we are choosing an option here